### DISCUSSION TITLE
Personalized/Adaptive Federated XGBoost

### DISCUSSION THREAD

**Discussion by WilliamLindskog** [NVIDIA/NVFlare#1724] [Ideas]
Dear NVFlare Team,

**Is your feature request related to a problem? Please describe.**
When dealing with Tabular Data, the Federated XGBoost that you provide is great! However, when the data are non-IID it would be great if you would provide a description of techniques, or a modified Federated XGBoost to deal with such issues. 

**Describe the solution you'd like**
Either:  (1) ways to improve local performance using personalized/adaptive approach, or (2) a completely personalized federated XGBoost model. For the former, could you please include what features to tune and how?

**Describe alternatives you've considered**
Sampling by important features (locally), gradient-based sampling, "adaptation" parameter. 

**Additional context**
Similar efforts can be found in *[Adaptive histogram-based gradient boosted trees for federated learning](https://arxiv.org/abs/2012.06670)* and *[Federated XGBoost on Sample-Wise Non-IID Data](https://arxiv.org/abs/2209.01340)*

Best regards,
W

**Comment by chesterxgchen** [comment]
thanks for raising the issue, we will look into this.

**Comment by ZiyueXu77** [comment]
Hi @WilliamLindskog, thanks for sharing this! I read the papers and one question regarding "non-iid" is that it seems the simulated non-iid is on sample size, if that's the case, we currently have some results with different data splits in our examples, with some empirical solutions to improve the results (weighted local learning rate for tree-based method).

Meanwhile if I understand correctly, the paper appears to be a histogram-based collaboration (sharing gradients and hessians), I wonder how we can simulate the shift in data distribution beyond the sample size in that sense. Please point out any misunderstandings, and it will be great if you have an actual non-iid dataset to do experiments on.

**Comment by WilliamLindskog** [comment]
Hi @ZiyueXu77, thanks for your comments! 

Yes, it would seem that the authors construct non-IID partitions mainly based on number of samples allocated to arbitrary client. Even though there aren't too many federated tabular datasets, one might be [FEMNIST](https://leaf.cmu.edu/). 

I suppose weighted local learning rate could be a possibility, other (still initial state) ideas I have thought about are: 

- Adaptive column sampling, putting more emphasis on specific columns for clients. 
- Global vs. local parts of the tree, similar to __parameter decoupling__ discussed in _Towards Personalized Federated Learning, Tan, AZ (2022)_ 

These are just some rough ideas I thought about. 
/W

**Comment by ZiyueXu77** [comment]
Hi @WilliamLindskog, thanks for the ideas! Let us know if you have further concrete specific needs during your implementation, and welcome to contribute to this project in the future. I will close this issue for now.

**Comment by YuanTingHsieh** [comment]
@WilliamLindskog I move this to discussion as issues are for more specific needs/requirements as @ZiyueXu77 mentioned.
 thanks
